<!DOCTYPE html>
<html lang="en">

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="">
    <meta name="author" content="">
    <!-- <base href='/clustergrammer/'> -->

    <title>Clustergrammer</title>

    <link rel="shortcut icon" href="static/icons/graham_cracker.ico"  type="image/x-icon">

    <link rel="stylesheet" href="lib/css/bootstrap.css">
    <link rel="stylesheet" href="lib/css/jquery-ui.css">
    <link rel="stylesheet" href="lib/font-awesome.min.css">

    <link rel="stylesheet" href="css/custom_multiple_clust.css">

    <style>
      #input_form {display: none}
    </style>

</head>

<body>

  <div id="wrap" class='toggled container'>

    <div class='row page_title' id='clustergram_title'>
      CST Lung Cancer Post-translational Modification Analysis
    </div>

    <div class='viz_description' id='clustergram_desc'>
      <p>Lung cancer is a complex disease that is known to be regulated at the level of post-translational modification, e.g. phosphorylation by driver kinases. Researchers at Cell Signaling Technology used SILAC mass spectrometry to measure differential phosphorylation, acetylation, and methylation in a panel of 42 lung cancer cell lines compared to non-cancerous tissue. Gene expression data from 37 of these lung cancer cell lines was obtained from the publically available Cancer Cell Line Encyclopedia (CCLE). </p>
    </div>

    <div class='viz_description' id='clustergram_desc'>
      <p>Below we use the web-based visualization tool, Clustergrammer, to overview this data. The lung cancer cell lines are shown as columns and prior knowledge, e.g. histological categorization, is shown as categories. Post-translational-modifications (PTMs) and expressed genes are shown as rows. The visualizations can be interatively explored and enrichment analysis can be performed using Enrichr.</p>
    </div>

    <div class='row viz_title' id='clustergram_title'>
      Lung Cancer Phosphorylation
    </div>

    <div class="row" >
      <div id='container-id-1' class='clustergrammer_container'>
          <div class='wait_message'>Please wait ...</div>
      </div>
    </div>

    <div class='viz_description' id='clustergram_desc'>
      <p> This is a clustergram visualization of differential phosphorylation from 42 lung cancer cell lines. The cell lines and phosphorylations have been hierarchically clustered using the Scipy library in Python using cosine distance and average linkage. The data has been quantile normalized across the cell lines (to make cell-line distributions similar) and zscored across the rows to improve facilitate comparison of differential phosphorylation across the cell lines.
      </p>
    </div>

    <div class='viz_description' id='clustergram_desc'>
      <p>Cell lines have been assigned categories based on their histology and their grouping based on gene expression data (exp-group-1, etc.). Note that cell lines display similar clustering based on phosphorylation and gene-expression data.
      </p>
    </div>


    <div class='row viz_title' id='col_sim_title'>
      Lung Cancer Gene Expression
    </div>

    <div class="row" >
      <div id='container-id-2' class='clustergrammer_container sim_mat_container'>
          <div class='wait_message'>Please wait ...</div>
      </div>
    </div>

    <div class='viz_description' id='col_sim_desc'>
      <p>This is a clustergram visualization of differential gene expression data from 37 lung cancer cell lines obtained from the CCLE. The cell lines adn genes have been hiararchically clustered using the Scipy library in Python using cosine distance and average linkage. Gene expression data was z-score normalized across all 1047 cell lines in the CCLE and only genes that have a zscore greater than 3 in more than 4 cell lines are shown.</p>
    </div>



<!--     <div class='row viz_title' id='row_sim_title'>
      Row Similarity Matrix
    </div>

    <div class="row" >
      <div id='container-id-3' class='clustergrammer_container sim_mat_container'>
          <div class='wait_message'>Please wait ...</div>
      </div>
    </div>

    <div class='viz_description' id='row_sim_desc'>
      <p>Above is a similarity matrix of the rows in your input matrix. The cells in the matrix represent the similarity between rows, where red/blue represent positive/negative similarity (measured as 1 - cosine-distance). </p>
    </div> -->


  </div>

  <div id='footer'>
    <div class="row" >
      <div id='footer_text_container' class="col-xs-12 footer_section">
        <div class="text-muted footer_text">Clustergrammer is being developed by the <a class='blue_links' target="_blank"  href="http://icahn.mssm.edu/research/labs/maayan-laboratory">Ma'ayan Lab</a> at the <a class='blue_links' target="_blank"   href="http://icahn.mssm.edu/">Icahn School of Medicine at Mount Sinai</a>  for the <a target="_blank" href="http://lincs-dcic.org/">BD2K-LINCS DCIC</a> and the <a target="_blank" href="http://commonfund.nih.gov/idg/overview">KMC-IDG</a>  </div>
        <!-- DCIC, BD2K, KMC, and GitHub -->
        <div class="text-muted footer_text">
        and is an open source project available on GitHub: <a class='blue_links' target="_blank"  href="https://github.com/MaayanLab/clustergrammer">Clustergrammer</a> and <a class='blue_links' target="_blank"  href="https://github.com/cornhundred/clustergrammer.js">Clustergrammer.js</a>

        </div>

      </div>
    </div>
  </div>


  <!-- Required JS Libraries -->
  <script src="lib/js/d3.js"></script>
  <script src="lib/js/jquery-1.11.2.min.js"></script>
  <script src="lib/js/jquery-ui.js"></script>
  <script src="lib/js/bootstrap.min.js"></script>
  <script src="lib/js/underscore-min.js"></script>
  <script src="lib/js/blockUI.js"></script>

  <!-- Clustergrammer JS -->
  <script src='js/clustergrammer.js'></script>
  <script src='js/enrichr_functions.js'></script>
  <script src='js/load_homepage.js'></script>

  <script type="text/javascript">

    load_viz_data();

  </script>

</body>

